### ISSUE TITLE
EF Core treats generated and directly-written projection expressions differnetly (When called from 'VisitLambda', rewriting a node of type... exception)

### ISSUE THREAD

**Issue by aradalvand** [dotnet/efcore#22943]
I'm facing a very strange case and I've been beating my head against the wall for quite a few hours and I can't make sense of it.

The problem: Passing an expression that was generated in runtime by an expression visitor to the `Select` method would cause EF Core to throw an exception, while passing the **EXACT SAME** expression but this time written directly in the code would work perfectly.
Take a look at the below code, it's self-explanatory:

```csharp
Expression<Func<Book, BookDto>> originalExpr = b => b.ToBookDto();

Expression<Func<Book, BookDto>> generatedExpr = new ExpandableVisitor().Visit(originalExpr) as Expression<Func<Book, BookDto>>;
Expression<Func<Book, BookDto>> sameAsGeneratedExpr = b => new BookDto()
{
    Id = b.Id,
    Name = b.Name,
    Author = new AuthorDto()
    {
        Id = b.Author.Id,
        Name = b.Author.Name,
        Books = b.Author.Books.Select(b => new BookDto()
        {
            Id = b.Id,
            Name = b.Name,
            Author = new AuthorDto()
            {
                Id = b.Author.Id,
                Name = b.Author.Name,
                Books = b.Author.Books.Select(b => b.ToBookDto()).ToList()
            }
        }).ToList()
    }
};

bool areExpressionsTheSame = generatedExpr.ToString() == sameAsGeneratedExpr.ToString(); // This is TRUE, which shows that there is NO DIFFERENCE between the 2 expressions.

IQueryable<BookDto> queryable = dbContext.Books.Select(generatedExpr); // Throws an exception
//IQueryable<BookDto> queryable = dbContext.Books.Select(sameAsGeneratedExpr); // Works properly

var dtos = queryable.Select(b => new BookDto()
{
    Name = b.Name,
    Author = new AuthorDto()
    {
        Name = b.Author.Name,
        Books = b.Author.Books.Select(b1 => new BookDto()
        {
            Name = b1.Name
        }).ToList()
    }
}).ToList(); // Note that this is the line that throws the exception, if that matters.
```
Using `generatedExpr` would cause the following exception to be thrown:
> "When called from 'VisitLambda', rewriting a node of type 'System.Linq.Expressions.ParameterExpression' must return a non-null value of the same type. Alternatively, override 'VisitLambda' and change it to not visit children of this type."

I have run out of ideas about what's causing this, the expressions are the same, how could EF Core even tell the difference?!
I can also create a public GitHub repo that you can clone to reproduce the problem if that's what you need, I couldn't put the whole codebase in here.

### Provider and version information:
EF Core version: 5.0.0 RC1
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 5.0

**Comment by smitpatel** [comment]
TLDR: Used wrong parameter expression reference. 2 parameter expressions which has same name and type would show up as same in string output but if they are not same objects then the expressions are different.

EF Core has ExpressionEqualityComparer which you can use if you really want to verify if the expressions are identical.

Long version (which is shorter) - Without the code which is generate the expression tree, there is hardly anything we can do here.

**Comment by AradAral** [comment]
@smitpatel Thanks for your response.
When I use the `ExpressionEqualityComparer` between `generatedExpr` and `sameAsGeneratedExpr`, I get this exception: `System.ArgumentException: 'An item with the same key has already been added. Key: b'`.

**Comment by AradAral** [comment]
I changed the parameter name b in the sameAsGeneratedExpr to book, and error has gone, which is weird because I thought generatedExpr is causing this exception.

Now when I check the equality:
```csharp
bool areTheSame = ExpressionEqualityComparer.Instance.Equals(sameAsGeneratedExpr, generatedExpr);
```
This will be true.

**Comment by smitpatel** [comment]
```
Expression<Func<Book, BookDto>> sameAsGeneratedExpr = b => new BookDto()
{
    Id = b.Id,
    Name = b.Name,
    Author = new AuthorDto()
    {
        Id = b.Author.Id,
        Name = b.Author.Name,
        Books = b.Author.Books.Select(b => new BookDto()
        {
            Id = b.Id,
            Name = b.Name,
            Author = new AuthorDto()
            {
                Id = b.Author.Id,
                Name = b.Author.Name,
                Books = b.Author.Books.Select(b => b.ToBookDto()).ToList()
            }
        }).ToList()
    }
};
```
The top level select, Books.Select, b.Author.Books.Select use same parameter name `b`. Compiler would pick up most nearest definition likely. While generating this tree manually, I believe there is error in declaring new parameter expression for each one.

**Comment by AradAral** [comment]
@smitpatel I modified the expression visitor such that it uses unique names for the parameters of each new lambda within the expression, and the EF exception is now no longer thrown, and the problem is apparently solved.
Although I'm still not completely sure what caused it in the first place.
Thank you for your help, sir.

**Comment by smitpatel** [comment]
@AradAral - Glad it is working now.

Without looking at code which is generating the dynamic tree, I can't say what piece actually threw off EF translation. But the root cause is that, EF Core deals with parameter expression using references so that the correct parameter is replaced when translating. Further expression equality comparer threw error about same parameter expression (again by reference) being used multiple times. Which points in the direction that a parameter expression was being used in incorrect place for some lambda.

While compiler worked fine for manually written lambdas too, in generally it is somewhat ambiguous to have same variable name for lambda parameter as it would be hard to figure out by inspecting which value would be used unless you know compiler rules. So one way to fix it is to generate new names for parameter every time. Or you can also use `Expression.Parameter` without name and compiler will take care of assigning it a unique name.
